pkg/extensions/extension_scrub.go

```diff
@@ -4,30 +4,107 @@
 package extensions
 
 import (
+	"errors"
+	"io"
+	goSync "sync"
 	"time"
 
 	"zotregistry.io/zot/pkg/api/config"
+	"zotregistry.io/zot/pkg/extensions/scheduler"
 	"zotregistry.io/zot/pkg/extensions/scrub"
 	"zotregistry.io/zot/pkg/log"
 	"zotregistry.io/zot/pkg/storage"
 )
 
 // EnableScrubExtension enables scrub extension.
-func EnableScrubExtension(config *config.Config, log log.Logger, run bool, imgStore storage.ImageStore, repo string) {
-	if !run {
-		if config.Extensions.Scrub != nil &&
-			config.Extensions.Scrub.Interval != 0 {
-			minScrubInterval, _ := time.ParseDuration("2h")
+func EnableScrubExtension(config *config.Config, log log.Logger, storeController storage.StoreController,
+	sch *scheduler.TaskScheduler,
+) {
+	if config.Extensions.Scrub != nil &&
+		config.Extensions.Scrub.Interval != 0 {
+		minScrubInterval, _ := time.ParseDuration("2h")
 
-			if config.Extensions.Scrub.Interval < minScrubInterval {
-				config.Extensions.Scrub.Interval = minScrubInterval
+		if config.Extensions.Scrub.Interval < minScrubInterval {
+			config.Extensions.Scrub.Interval = minScrubInterval
 
-				log.Warn().Msg("Scrub interval set to too-short interval < 2h, changing scrub duration to 2 hours and continuing.") //nolint:lll // gofumpt conflicts with lll
+			log.Warn().Msg("Scrub interval set to too-short interval < 2h, changing scrub duration to 2 hours and continuing.") //nolint:lll // gofumpt conflicts with lll
+		}
+
+		var periodicScrub *periodicScrubTask
+
+		periodicScrub = newPeriodicScrubTask(config.Extensions.Scrub.Interval, storeController.DefaultStore, log, sch)
+		sch.AddTask(periodicScrub)
+
+		if config.Storage.SubPaths != nil {
+			for route := range config.Storage.SubPaths {
+				periodicScrub = newPeriodicScrubTask(config.Extensions.Scrub.Interval, storeController.SubStore[route], log, sch)
+				sch.AddTask(periodicScrub)
 			}
-		} else {
-			log.Info().Msg("Scrub config not provided, skipping scrub")
 		}
 	} else {
-		scrub.RunScrubRepo(imgStore, repo, log)
+		log.Info().Msg("Scrub config not provided, skipping scrub")
 	}
 }
+
+type periodicScrubTask struct {
+	*scheduler.AbstractTask
+	interval time.Duration
+	imgStore storage.ImageStore
+	log      log.Logger
+	sch      *scheduler.TaskScheduler
+	cancel   chan bool
+}
+
+func newPeriodicScrubTask(interval time.Duration, imgStore storage.ImageStore, log log.Logger,
+	sch *scheduler.TaskScheduler,
+) *periodicScrubTask {
+	abstractTask := &scheduler.AbstractTask{}
+	cancel := make(chan bool)
+	periodicScrub := &periodicScrubTask{abstractTask, interval, imgStore, log, sch, cancel}
+	abstractTask.Task = periodicScrub
+
+	return periodicScrub
+}
+
+func (periodicScrubT *periodicScrubTask) DoWork(wgScheduler *goSync.WaitGroup) (scheduler.Status, scheduler.Cb) {
+	ticker := time.NewTicker(periodicScrubT.interval)
+
+	go func() {
+		for {
+			select {
+			case <-periodicScrubT.cancel:
```

Comment:
Why not use a context - also for handling timeouts and cancellation - see some examples here https://www.sohamkamani.com/golang/context-cancellation-and-values/